server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -1514,7 +1514,7 @@ public boolean isSegmentReplicationAllowed() {
             return false;
         }
         if (state().equals(IndexShardState.STARTED) == false
-            && (state() == IndexShardState.POST_RECOVERY && shardRouting.state() == ShardRoutingState.INITIALIZING) == false) {
+            && (state() == IndexShardState.RECOVERING && shardRouting.state() == ShardRoutingState.INITIALIZING) == false) {
```

Comment:
when we do a round of segrep from recovery it's usually from finalizerecovery() which happens after all recovery process is finished, which is fine. But what if on a refresh from primary, replica receives a checkpoint and it's still in recovery (not reached finalize recovery yet), then replica shard will also start segment replication while recovery is still in progress. Are we fine if we somehow end up in this state?

Reply:
>But what if on a refresh from primary, replica receives a checkpoint and it's still in recovery (not reached finalize recovery yet), then replica shard will also start segment replication while recovery is still in progress.

We should prevent segment replication before/during recovery. This should already be handled on target side. Please let me know if there is use case where it is possible.